<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulation of K-mer Data Based on Protein Sequences • kmerFilters</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Simulation of K-mer Data Based on Protein Sequences">
<meta property="og:description" content="This package is designed for the simulation of k-mer data. It generates synthetic k-mer sequences, which are crucial for various bioinformatics analyses. The package allows users to customize parameters to mimic different biological scenarios. It is especially useful for testing algorithms and validating computational models in proteomics.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">kmerFilters</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/about_kmers.html">Data sampling and data structure.</a>
    </li>
    <li>
      <a href="articles/simulation_description.html">Motif injection and sequences' labels.</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="kmerfilters--r-package-for-k-mer-data-simulation-and-filtering">kmerFilters : R Package for k-mer Data Simulation and Filtering<a class="anchor" aria-label="anchor" href="#kmerfilters--r-package-for-k-mer-data-simulation-and-filtering"></a>
</h1></div>
<!-- badges: start -->

<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>k-mers (n-grams) refer to k-length substrings derived from longer sequences, which can be continuous (representing a block of subsequent residues) or discontinuous (where the wildcards representing the gaps between residues are allowed). In biological applications, k-mers are used for various purposes, including genome assembly, sequence alignment, motif discovery, variant calling, phylogenetics, and CRISPR target identification. Due to the vast number of variables introduced by k-mer notation, we need tools to filter the variables.</p>
<p>In this package we provide tools for simulating k-mer data and benchmarking various filtering techniques. It is designed to assist researchers and practitioners in evaluating and comparing different approaches for preprocessing sequence data, particularly for applications such as protein function prediction.</p>
</div>
<div class="section level2">
<h2 id="features">Features<a class="anchor" aria-label="anchor" href="#features"></a>
</h2>
<p>Our package provides fast and efficient generation of synthetic k-mer datasets with customizable parameters, including sequence length, number of sequences, k-mer size and motifs impact. The latest embraces modelling the response variable in discrete or continuous way depending on few biological cases, such as</p>
<ul>
<li><p>additive impact of each motif,</p></li>
<li><p>additive impact of interactions of motifs,</p></li>
<li><p>additive impact of custom interactions (logical expressions using motifs).</p></li>
</ul>
<p>To learn more, please visit: <a href="https://biogenies.info/kmerFilters/" class="external-link uri">https://biogenies.info/kmerFilters/</a></p>
<p>Specific resources:</p>
<ul>
<li><a href="https://biogenies.info/kmerFilters/articles/simulation_description.html" class="external-link">Simulation description</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="how-to-install">How to install<a class="anchor" aria-label="anchor" href="#how-to-install"></a>
</h2>
<p>You can install the latest version from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># install.packages("devtools")</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"BioGenies/kmerFilters"</span>)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="examples-of-usage">Examples of usage<a class="anchor" aria-label="anchor" href="#examples-of-usage"></a>
</h2>
<div class="section level3">
<h3 id="how-to-simulate-data">How to simulate data<a class="anchor" aria-label="anchor" href="#how-to-simulate-data"></a>
</h3>
<p>You can use the <code>generate_motifs</code> function for motifs generation. For example:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">kmerFilters</span><span class="op">)</span></span>
<span></span>
<span><span class="va">alph</span> <span class="op">&lt;-</span> <span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="va">n_injections</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span></span>
<span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generate_motifs.html">generate_motifs</a></span><span class="op">(</span>alphabet <span class="op">=</span> <span class="va">alph</span>, </span>
<span>                          n_motifs <span class="op">=</span> <span class="fl">4</span>, </span>
<span>                          n_injections <span class="op">=</span> <span class="fl">4</span>, </span>
<span>                          n <span class="op">=</span> <span class="fl">4</span>, </span>
<span>                          d <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="va">motifs</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "d" "_" "_" "_" "_" "_" "b"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] "c" "_" "_" "d" "c"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [1] "b" "_" "c" "_" "_" "c"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; [1] "c" "_" "_" "b" "_" "_" "c"</span></span></code></pre></div>
<p>Using simulated motifs we can simulate positive and negative sequences and consrtuct a k-mer feature space:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generate_kmer_data.html">generate_kmer_data</a></span><span class="op">(</span>n_seq <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                              sequence_length <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                              alphabet <span class="op">=</span> <span class="va">alph</span>,</span>
<span>                              motifs <span class="op">=</span> <span class="va">motifs</span>, </span>
<span>                              n_injections <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">results</span></span>
<span><span class="co">#&gt; 20 x 14940 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;   [[ suppressing 33 column names 'a', 'd', 'b' ... ]]</span></span>
<span><span class="co">#&gt;                                                                               </span></span>
<span><span class="co">#&gt;  [1,] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 . . . . . . 1 1 1 1 1 1 1 1 1 1 1 1 . ......</span></span>
<span><span class="co">#&gt;  [2,] 1 1 1 1 . 1 1 1 1 1 1 1 1 1 1 1 1 . . . 1 1 . 1 1 1 1 1 . 1 . 1 1 ......</span></span>
<span><span class="co">#&gt;  [3,] 1 1 1 1 . 1 . 1 . . 1 1 1 1 1 1 1 1 1 1 1 1 1 1 . 1 1 1 1 . 1 1 1 ......</span></span>
<span><span class="co">#&gt;  [4,] 1 1 1 1 . 1 1 1 1 1 . 1 1 . 1 1 1 1 1 1 . 1 1 1 . 1 . 1 1 1 1 1 1 ......</span></span>
<span><span class="co">#&gt;  [5,] 1 1 1 1 1 1 1 . 1 1 1 1 1 1 . 1 1 1 . . 1 1 1 . 1 1 . 1 1 1 1 1 . ......</span></span>
<span><span class="co">#&gt;  [6,] 1 1 1 1 1 . 1 1 1 1 1 1 . 1 . 1 . 1 1 . 1 1 . 1 . . . 1 1 1 1 . 1 ......</span></span>
<span><span class="co">#&gt;  [7,] 1 1 1 1 1 1 . 1 1 . 1 1 1 1 1 1 . 1 1 1 . 1 1 . 1 1 . 1 1 1 1 . . ......</span></span>
<span><span class="co">#&gt;  [8,] 1 1 1 1 1 . 1 . 1 1 1 1 . 1 . 1 1 . . . 1 . . 1 1 . 1 1 . 1 1 1 . ......</span></span>
<span><span class="co">#&gt;  [9,] 1 1 1 1 1 1 . 1 1 1 1 1 1 1 . 1 . . 1 1 1 . . 1 1 . 1 1 1 1 1 1 1 ......</span></span>
<span><span class="co">#&gt; [10,] 1 1 1 1 1 1 1 1 . 1 1 1 1 . . . 1 1 . . 1 1 1 . 1 1 1 1 . 1 1 1 . ......</span></span>
<span><span class="co">#&gt; [11,] 1 1 1 1 . . . 1 . 1 . 1 1 1 1 1 1 1 1 . 1 1 1 1 1 1 1 1 1 . 1 . . ......</span></span>
<span><span class="co">#&gt; [12,] 1 1 1 1 1 1 1 . . 1 1 1 1 1 1 1 1 . 1 . 1 1 . . 1 1 1 1 . 1 1 1 . ......</span></span>
<span><span class="co">#&gt; [13,] 1 1 1 1 1 1 1 . 1 1 . 1 . 1 1 . 1 1 1 1 1 1 . . . 1 . 1 1 1 1 1 1 ......</span></span>
<span><span class="co">#&gt; [14,] 1 1 1 1 1 . 1 1 1 . 1 1 1 1 1 1 . . 1 1 1 1 1 1 . 1 1 1 1 1 . 1 . ......</span></span>
<span><span class="co">#&gt; [15,] 1 1 . 1 1 1 . 1 . 1 1 . 1 . . . 1 . 1 1 1 . 1 . 1 . 1 . 1 1 . 1 1 ......</span></span>
<span><span class="co">#&gt; [16,] 1 1 1 1 . 1 1 . 1 1 . 1 . 1 . 1 1 1 1 . 1 . 1 . . 1 . 1 . 1 1 . . ......</span></span>
<span><span class="co">#&gt; [17,] 1 1 1 1 1 1 1 1 1 1 1 1 . 1 1 . . 1 1 . 1 1 1 1 1 1 . . 1 1 1 1 1 ......</span></span>
<span><span class="co">#&gt; [18,] 1 1 1 1 1 1 1 1 1 1 . 1 1 . 1 . . 1 1 1 1 1 . 1 . 1 . 1 1 1 . 1 1 ......</span></span>
<span><span class="co">#&gt; [19,] 1 1 1 1 . . 1 1 . 1 . 1 1 1 1 1 1 1 1 1 . 1 . 1 . . . 1 1 1 1 . 1 ......</span></span>
<span><span class="co">#&gt; [20,] 1 1 1 1 1 . 1 1 1 . 1 1 1 1 1 1 . 1 1 1 1 1 . . . 1 1 1 1 1 1 . 1 ......</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  .....suppressing 14907 columns in show(); maybe adjust 'options(max.print= *, width = *)'</span></span>
<span><span class="co">#&gt;  ..............................</span></span></code></pre></div>
<p>Using obtained data you can choose how to generate a response variable. We provide three functions for that:</p>
<ul>
<li>
<code><a href="reference/get_target_additive.html">get_target_additive()</a></code>,</li>
<li>
<code><a href="reference/get_target_interactions.html">get_target_interactions()</a></code>,</li>
<li>
<code><a href="reference/get_target_logic.html">get_target_logic()</a></code>.</li>
</ul>
<p>For example, the following code:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/get_target_additive.html">get_target_additive</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] 0 1 1 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0</span></span></code></pre></div>
<p>creates a binary response variable based on the logistic regression model assumptions.</p>
</div>
<div class="section level3">
<h3 id="how-to-filter">How to filter<a class="anchor" aria-label="anchor" href="#how-to-filter"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h2>
<p>Contributions to kmerFilters are welcome! If you have suggestions for new features, improvements, or bug fixes, please submit an issue or pull request on GitHub.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 3)</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing kmerFilters</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Michał Burdukiewicz <br><small class="roles"> Author, contributor </small> <a href="https://orcid.org/0000-0001-8926-582X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Krystyna Grzesiak <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-2581-7722" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jakub Kała <br><small class="roles"> Author, contributor </small> <a href="https://orcid.org/0000-0002-7187-6988" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://app.codecov.io/gh/BioGenies/kmerFilters?branch=main" class="external-link"><img src="https://codecov.io/gh/BioGenies/kmerFilters/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Michał Burdukiewicz, Krystyna Grzesiak, Jakub Kała.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
