<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data sampling and data structure. ‚Ä¢ kmerFilters</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Data sampling and data structure.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">kmerFilters</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/about_kmers.html">Data sampling and data structure.</a>
    </li>
    <li>
      <a href="../articles/simulation_description.html">Motif injection and sequences' labels.</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Data sampling and data structure.</h1>
            
      

      <div class="hidden name"><code>about_kmers.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="what-are-k-mers">What are k-mers?<a class="anchor" aria-label="anchor" href="#what-are-k-mers"></a>
</h2>
<p>k-mers, or n-grams, are sequences of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
consecutive elements extracted from a larger sequence of letters. In the
context of bioinformatics, k-mers refer to short DNA, RNA, or protein
sequences. They are used to analyze biological sequences by breaking
them down into manageable fragments, allowing researchers to identify
patterns, mutations, and other genetic features. k-mers are valuable in
various applications, including sequence alignment, genome assembly, and
gene prediction.</p>
<p>In our package we generate some artificial sequences and count k-mers
for them using the <a href="https://github.com/slowikj/seqR" class="external-link">seqr</a>
package. If you came here with your own data for analysis and do not
wish to simulate artificial data, you can use the <a href="https://github.com/slowikj/seqR" class="external-link">seqr</a> package to count k-mers
in your sequences. Anyway, you can filter k-mers with methods contained
in <em>kmerFilters</em>.</p>
</div>
<div class="section level2">
<h2 id="data-structure">Data structure<a class="anchor" aria-label="anchor" href="#data-structure"></a>
</h2>
<p>To generate data we first generate both positive and negative
sequences and then count k-mers for them. Everything is done by two
functions specified by the following parameters (for more see the
documentation):</p>
<ul>
<li>alphabet - set of letters for sampling,</li>
<li>sequence_length - length of sequence,</li>
<li>n_injections - maximum number of motifs to inject into a positive
sequence.</li>
</ul>
<p>Let‚Äôs create an example set of motifs</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">111</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dna_alphabet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"T"</span>, <span class="st">"G"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_motifs.html">generate_motifs</a></span><span class="op">(</span>alphabet <span class="op">=</span> <span class="va">dna_alphabet</span>,     <span class="co"># alphabet</span></span>
<span>                          n_motifs <span class="op">=</span> <span class="fl">2</span>,                <span class="co"># number of motifs</span></span>
<span>                          n_injections <span class="op">=</span> <span class="fl">1</span>,            <span class="co"># number of injections</span></span>
<span>                          n <span class="op">=</span> <span class="fl">3</span>,                       <span class="co"># number of letters</span></span>
<span>                          d <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>                       <span class="co"># number of possible gaps</span></span></code></pre></div>
<p>Here, the result is a list of motifs for potential injection to
positive sequences:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">motifs</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] "T" "G" "T"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] "A" "_" "T"</span></span></code></pre>
<p>Having motifs we can generate sequences and create a k-mer feature
space:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequence_length</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span></span>
<span><span class="va">kmer_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_kmer_data.html">generate_kmer_data</a></span><span class="op">(</span>n_seq <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                               sequence_length <span class="op">=</span> <span class="va">sequence_length</span>, </span>
<span>                               alph <span class="op">=</span> <span class="va">dna_alphabet</span>,</span>
<span>                               motifs <span class="op">=</span> <span class="va">motifs</span>, </span>
<span>                               n_injections <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                               n <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                               d <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>The output of <code>generate_kmer_data</code> function is an object
of the <code>dgCMatrix</code> class. It‚Äôs a sparse matrix containing 0‚Äôs
and 1‚Äôs where 1 denotes the occurrence of k-mer and 0 denotes no
occurrence. The below numbers means that we obtained 10 rows (sequences)
and 28 (found k-mers with maximum 2 letetrs and 1 gap):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">kmer_dat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10 28</span></span></code></pre>
<p>We can easily access the k-mers using the following command:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kmer_dat</span><span class="op">@</span><span class="va">Dimnames</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "G_0"   "T_0"   "C_0"   "A_0"   "T.G_0" "G.T_0" "T.T_0" "T.A_0" "A.C_0"</span></span>
<span><span class="co">## [10] "C.T_0" "C.G_0" "G.C_0" "G.G_0" "G.A_0" "A.A_0" "T.C_0" "A.T_0" "T.G_1"</span></span>
<span><span class="co">## [19] "T.T_1" "G.G_1" "A.T_1" "C.A_1" "G.A_1" "G.T_1" "C.C_1" "C.G_1" "T.A_1"</span></span>
<span><span class="co">## [28] "T.C_1"</span></span></code></pre>
<p>Each dot <code>.</code> between letters represents a potential gap.
The numbers following the <code>_</code> symbol indicate the number of
gaps associated with each dot. For example, <code>A.C_1</code> means
exactly
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtext mathvariant="monospace">ùô∞_ùô≤</mtext><annotation encoding="application/x-tex">\texttt{A_C}</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtext mathvariant="monospace">_</mtext><annotation encoding="application/x-tex">\texttt{_}</annotation></semantics></math>
means a gap. You can decode the names using <a href="https://github.com/michbur/biogram" class="external-link">biogram</a> package as
follows</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="fu">biogram</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biogram/man/decode_ngrams.html" class="external-link">decode_ngrams</a></span><span class="op">(</span><span class="va">kmer_dat</span><span class="op">@</span><span class="va">Dimnames</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "G"   "T"   "C"   "A"   "TG"  "GT"  "TT"  "TA"  "AC"  "CT"  "CG"  "GC" </span></span>
<span><span class="co">## [13] "GG"  "GA"  "AA"  "TC"  "AT"  "T_G" "T_T" "G_G" "A_T" "C_A" "G_A" "G_T"</span></span>
<span><span class="co">## [25] "C_C" "C_G" "T_A" "T_C"</span></span></code></pre>
<p>For a set of sequences each k-mer is a variable indicating whether
given k-mer occurs in corresponding sequence. Here, we have 28
k-mers:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">kmer_dat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       G_0 T_0 C_0 A_0 T.G_0 G.T_0 T.T_0 T.A_0 A.C_0 C.T_0 C.G_0 G.C_0 G.G_0</span></span>
<span><span class="co">##  [1,]   1   1   0   0     1     1     1     0     0     0     0     0     0</span></span>
<span><span class="co">##  [2,]   1   1   0   0     1     1     0     0     0     0     0     0     0</span></span>
<span><span class="co">##  [3,]   1   1   0   0     1     1     0     0     0     0     0     0     0</span></span>
<span><span class="co">##  [4,]   0   1   1   1     0     0     0     1     1     1     0     0     0</span></span>
<span><span class="co">##  [5,]   1   1   0   1     1     1     0     1     0     0     0     0     0</span></span>
<span><span class="co">##  [6,]   1   1   1   0     0     0     0     0     0     1     1     1     0</span></span>
<span><span class="co">##  [7,]   1   0   0   1     0     0     0     0     0     0     0     0     1</span></span>
<span><span class="co">##  [8,]   1   1   1   0     1     0     0     0     0     1     0     0     0</span></span>
<span><span class="co">##  [9,]   1   1   0   1     0     0     0     0     0     0     0     0     0</span></span>
<span><span class="co">## [10,]   0   1   1   1     0     0     1     1     1     0     0     0     0</span></span>
<span><span class="co">##       G.A_0 A.A_0 T.C_0 A.T_0 T.G_1 T.T_1 G.G_1 A.T_1 C.A_1 G.A_1 G.T_1 C.C_1</span></span>
<span><span class="co">##  [1,]     0     0     0     0     1     1     0     0     0     0     0     0</span></span>
<span><span class="co">##  [2,]     0     0     0     0     0     1     1     0     0     0     0     0</span></span>
<span><span class="co">##  [3,]     0     0     0     0     0     1     1     0     0     0     0     0</span></span>
<span><span class="co">##  [4,]     0     0     0     0     0     0     0     1     1     0     0     0</span></span>
<span><span class="co">##  [5,]     0     0     0     0     0     1     0     0     0     1     0     0</span></span>
<span><span class="co">##  [6,]     0     0     0     0     0     0     0     0     0     0     1     1</span></span>
<span><span class="co">##  [7,]     1     1     0     0     0     0     0     0     0     1     0     0</span></span>
<span><span class="co">##  [8,]     0     0     1     0     0     1     0     0     0     0     0     0</span></span>
<span><span class="co">##  [9,]     1     1     0     1     0     0     0     1     0     1     0     0</span></span>
<span><span class="co">## [10,]     0     0     0     0     0     0     0     0     0     0     0     0</span></span>
<span><span class="co">##       C.G_1 T.A_1 T.C_1</span></span>
<span><span class="co">##  [1,]     0     0     0</span></span>
<span><span class="co">##  [2,]     0     0     0</span></span>
<span><span class="co">##  [3,]     0     0     0</span></span>
<span><span class="co">##  [4,]     0     0     0</span></span>
<span><span class="co">##  [5,]     0     0     0</span></span>
<span><span class="co">##  [6,]     0     0     0</span></span>
<span><span class="co">##  [7,]     0     0     0</span></span>
<span><span class="co">##  [8,]     1     0     0</span></span>
<span><span class="co">##  [9,]     0     0     0</span></span>
<span><span class="co">## [10,]     0     1     1</span></span></code></pre>
<p>corresponding to the sequences:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">kmer_dat</span>, <span class="st">"sequences"</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">paste0</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "TTGT" "GTGT" "TGTG" "ACTA" "TGTA" "CGCT" "GGAA" "TCTG" "GAAT" "TTAC"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="target-variable">Target variable<a class="anchor" aria-label="anchor" href="#target-variable"></a>
</h2>
<p>Here we will shortly describe the response variable. For the set of
generated motifs</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">kmer_dat</span>, <span class="st">"motifs_set"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] "T" "G" "T"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] "A" "_" "T"</span></span></code></pre>
<p>we create an object <code>motifs_map</code> which denotes which motif
was injected to which sequence. In our example</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">kmer_dat</span>, <span class="st">"motifs_map"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       [,1] [,2]</span></span>
<span><span class="co">##  [1,]    1    0</span></span>
<span><span class="co">##  [2,]    1    0</span></span>
<span><span class="co">##  [3,]    1    0</span></span>
<span><span class="co">##  [4,]    0    1</span></span>
<span><span class="co">##  [5,]    1    0</span></span>
<span><span class="co">##  [6,]    0    0</span></span>
<span><span class="co">##  [7,]    0    0</span></span>
<span><span class="co">##  [8,]    0    0</span></span>
<span><span class="co">##  [9,]    0    0</span></span>
<span><span class="co">## [10,]    0    0</span></span></code></pre>
<p>we can see that the motif TGT occurred 4 times and A_T occurred 1
time. The <code>target</code> attribute says whether given sequence
contains at least one motif:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">kmer_dat</span>, <span class="st">"target"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE</span></span></code></pre>
<p>A variable <code>target</code> says whether a sequence is positive or
negative, however we might also consider some noise (false positively or
negatively labeled sequences). To do so, we use the functions:</p>
<ul>
<li>
<code>get_target_additive</code>,</li>
<li>
<code>get_target_interactions</code>,</li>
<li>
<code>get_target_logic</code>.</li>
</ul>
<p>For example</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_target_additive.html">get_target_additive</a></span><span class="op">(</span><span class="va">kmer_dat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 1 0 1 0 1 1 1 0 0 1</span></span></code></pre>
<p>For more details about target sampling see <a href="https://biogenies.info/kmerFilters/articles/simulation_description.html" class="external-link">vignette</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Micha≈Ç Burdukiewicz, Krystyna Grzesiak, Jakub Ka≈Ça.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
